Code/src/main/java/nl/utwente/viskell/haskell/type/TypeVar.java

```diff
@@ -79,23 +98,42 @@ private ConcreteType get() {
 
         /**
          * @throws IllegalStateException when is isPresent() is true
+         * @throws HaskellTypeError if this type variable is rigid
          * @param The new instance of this type.
          */
-        private void set(ConcreteType type) {
+        private void set(ConcreteType type) throws HaskellTypeError {
             if (this.type != null) {
                 throw new IllegalStateException("Type instance already set");
             }
             
+            if (this.isRigid) {
+                throw new HaskellTypeError("Can not unify a rigid type variable " + this.name + " with concrete type " + this.type.prettyPrint());
+            }
+            
             this.type = type;
         }
 
         /**
          * Deeply unifying all aspects of type instances
          * @param the other type instance.
-         * @throws HaskellTypeError if the combined constraint set of the typeva instances is not satisfiable.
+         * @throws HaskellTypeError if the combined constraint set of the typevar instances is not satisfiable.
+         * @throws HaskellTypeError if a rigid type variable gets unified with more constraints or another rigid type variable
          */
         private void unifyWith(TypeInstance other) throws HaskellTypeError {
+            if (this == other) {
```

Comment:
If both are rigid and isomorphic, they shouldn't produce an exception, right? A rigid ::a and ::b can live happily together.